### PR TITLE
Ds/pybindings 2d

### PR DESCRIPTION
**Pull request by dsteinmo**
This should be everything we need for a 2D python hyperbolic system solve. The sw2d.py guy is still a work-in-progress, but the C++ is at a good 'first draft' stage.

### CODE REVIEW

**Review by dsteinmo — COMMENTED** [review]
This seems to produce correct output in the vtu plots now using `python3 sw2d.py`. I cut a few corners (making some fields row major that used to be column major) so that they would copy over to python row-major arrays more 'natively', this might cost us some optimizations, but I don't think it's terrible.

File: src/VtkOutputter.cpp [hunk 1]
```diff
@@ -130,11 +134,11 @@ namespace blitzdg {
 
 	void VtkOutputter::writeFieldsToFiles(std::map<std::string, real_matrix_type>& fields, index_type tstep) {
 		for (auto kv : fields) {
```
Comment (k7miller) [inline src/VtkOutputter.cpp]: This loop creates a copy in `kv` of each element in `fields`. You may want to write it as
`for(const auto& kv : fields)`

File: src/VtkOutputter.cpp [hunk 2]
```diff
@@ -250,5 +254,37 @@ namespace blitzdg {
...
+	void VtkOutputter::writeFieldsToFiles_numpy(const boost::python::dict& fields, index_type tstep) {
+
+		auto fieldNames = std::vector<const char*>(boost::python::stl_input_iterator<const char *>(fields.keys()),
+			stl_input_iterator<const char *>() );
+
+		auto fieldArrays = std::vector<ndarray>(boost::python::stl_input_iterator<ndarray>(fields.values()),
+			stl_input_iterator<ndarray>()); 
+
+		index_type count = 0;
+		for (auto f = fieldNames.begin(); f != fieldNames.end(); ++f) {
```
Comment (k7miller) [inline src/VtkOutputter.cpp]: A range-based for-loop might be nicer here.

File: src/DGContext2D.cpp [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        Py_intptr_t shape[1] = {  NumFaces_*Nfp_*K_ };
+        ndarray result = zeros(1, shape, dtype::get_builtin<index_type>());
+        std::copy(vmapP_->begin(), vmapP_->end(), reinterpret_cast<index_type*>(result.get_data()));
+        return result;
+    }
+
+    dict DGContext2D::bcmap_numpy() const {
+        boost::python::dict bcDict;
+
+        for (auto t = bcHash_->begin(); t != bcHash_->end(); ++t) {
```
Comment (k7miller) [inline src/DGContext2D.cpp]: Try with a range-based for-loop.

File: src/DGContext2D.cpp [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+            auto key = t->first;
+            auto vec = t->second;
+
+            boost::python::list bcList;
+            for (auto n = vec.begin(); n != vec.end(); ++n) {
```
Comment (k7miller) [inline src/DGContext2D.cpp]: Try with a range-based for-loop.

**Review by k7miller — DISMISSED** [review]
Code looks good! Mostly just some formatting changes.

**Pull request merged**